R/cube-collapse-dimensions.R

```diff
@@ -0,0 +1,127 @@
+#' @rdname cube-computing
+#' @export
+setMethod("collapse.dimensions", "CrunchCube", function (x, margin=NULL) {
+    # ensure that we have sensible margin input
+    selecteds <- is.selectedDimension(x@dims)
+    check_margins(margin, selecteds)
+    
+    # translate from user-cube margins to real-cube margins and establish the
+    # two groups of margins: those to collapse and those to keep
+    margins_to_collapse <- user2real(margin, cube = x)
+    margins_to_keep <- setdiff(seq_along(x@dims), margins_to_collapse)
+    
+    # if there are any _items in the margin to collapse, be wary! Since there
+    # are (likely) multiple items in the array, each row of the dataset would
+    # have more than on item here. If we simply summed across them (like we do
+    # with other measures below) then we would inflate the number of responses.
+    collapsed_items <- endsWith(getDimTypes(x)[margins_to_collapse], "_items")
+    
+    # grab the old cube structure and over-write it with new subset data.
+    out <- x
+    
+    # iterate through measures, collapsing the dimension(s) specified.
+    out@arrays[] <- lapply(
+        out@arrays,
+        collapse_with, 
+        margins_to_keep = margins_to_keep, 
+        margins_to_collapse = margins_to_collapse, 
+        collapsed_items = collapsed_items)
+    
+    # we needed to include the selected dim of MRs in off_margins above, but
+    # when subsetting dimensions we definitely do not want them to be included.
+    out@dims <- out@dims[margins_to_keep]
+    out$query$dimensions <- out$query$dimensions[margins_to_keep]
```

Comment:
I suspect that munging things that aren't in `@dims` and `@arrays` is not worth the effort to try to get right (maybe you have covered everything here, but I'm skeptical). IIRC we don't bother with this in the `[` method. I wouldn't want to cause an error for a user in this section while trying to update them, nor would I want to bother maintaining this logic when the API evolves. 

I'm pretty certain that all `CrunchCube` methods only look at dims and arrays and never reach into the query result. And in practice, all you care is that `prop.table` works so that you can use this in `index.table`.

Reply:
Originally this was to ensure I was not fooling myself in believing I could totally reconstruct the univariate cube, but I agree — it's probably more work than necessary to maintain, so I've removed it and now just test that `@arrays` and `@dims` are equal in the tests.